rfcs/20220630-bit-exact-conv/README.md

```diff
@@ -0,0 +1,101 @@
+# RFC: Optional bit exact Convolution
+
+## Introduction
+
+Several cpu convolution primitives in the library may compute a non accurate
+result in case vnni is unavailable.
+An example case is the `jit_avx512_core_x8s8s32x_convolution_fwd_t` convolution
+primitive running with avx512_core ISA.
+
+Non accurate results may occur due to the following behaviors:
+
+1. Weights and inputs multiplication uses the VPMADDUBSW instruction. This
+   instruction multiplies two pairs of u8/s8 values and then accumulates the
+   two s16 products into one s16 with potential saturation.
+
+2. Weights are scaled with a factor of 0.5f. This scaling is then compensated
+   for with a factor of 2. However, for odd weights, this compensation gives
+   non accurate results, as the weights are integers.
+
+These behaviors are intentional, giving significantly better performance than
+the equivalent accurate computation.
+However, depending on the user, compromising accuracy may not be an option.
+Adding an optional bit exact implementation will allow the user to prioritize
+correctness over performance.
+
+## Proposal
+
+The proposal is to let the user choose a slower implementation with guaranteed
+accuracy.
+
+The link below demonstrates the changes in the primitive's implementation,
+allowing accurate results.
+It does not demonstrate any API changes needed to expose this option to the
+user, instead it uses a compile time define.
+
+https://github.com/oneapi-src/oneDNN/compare/master...maayaneh:oneDNN:master
```

Comment:
Maybe outside the scope of this RFC, but let me go ahead and comment on the implementation at this link. From an integration scalability perspective, I think it would be preferable to minimize the set of changes needed to support the bit-exact mode. So I wonder if something like the following might work instead:
```diff
--- a/src/cpu/x64/jit_avx512_core_x8s8s32x_1x1_conv_kernel.cpp
+++ b/src/cpu/x64/jit_avx512_core_x8s8s32x_1x1_conv_kernel.cpp
@@ -519,6 +519,21 @@ void _jit_avx512_core_x8s8s32x_1x1_conv_kernel<Vmm>::reduce_loop(
     auto compute = [=](Vmm vreg_acc, Vmm vreg_wei, Vmm vreg_src) {
         if (jcp.has_vnni) {
             vpdpbusd(vreg_acc, vreg_src, vreg_wei);
+        } else if (jcp.force_be) {
+            // use vmm_wei as an extra tmp register
+            // use vmask to hold even/odd byte masks
+            // process even values
+            kshiftlq(vmask, vmask, 0x1); // 0xaa...
+            vmovdqu8(vmm_wei, vreg_wei | vmask)
+            vpmaddubsw(vmm_tmp, vreg_src, vmm_wei);
+            vpmaddwd(vmm_tmp, vmm_tmp, vmm_one);
+            vpaddd(vreg_acc, vreg_acc, vmm_tmp);
+            // process odd values
+            kshiftrq(vmask, vmask, 0x1); // 0x55...
+            vmovdqu8(vmm_wei, vreg_wei | vmask)
+            vpmaddubsw(vmm_tmp, vreg_src, vmm_wei);
+            vpmaddwd(vmm_tmp, vmm_tmp, vmm_one);
+            vpaddd(vreg_acc, vreg_acc, vmm_tmp);
         } else {
             vpmaddubsw(vmm_tmp, vreg_src, vreg_wei);
             vpmaddwd(vmm_tmp, vmm_tmp, vmm_one);
```
This allows to keep the same weights layouts and same general logic everywhere, but replace the `vpmaddubsw+vpmaddwd+vpaddd` sequence with a bit-exact version. This can simplify integration into other optimized convolution implementations like `jit_brgemm_conv.cpp`.

Reply:
Hi @kwiersch, thank you for your input!
I uploaded a different branch following your suggestion, using the same weights layout:
https://github.com/oneapi-src/oneDNN/compare/master...maayaneh:oneDNN:be_preserve_wei_layout

The changes are indeed much more concise, but it seems to be slower than the alternative (using OIhw8i16o2i layout).
If you have any thoughts on accelerating this implementation we'll be very interested.

Reply:
I see. Yeah, I was hoping that the performance might be close enough to justify the simpler approach. But sounds like it's not close enough.

As for improving performance of this more concise method, the only thing I can think of is to try adding a second weights register and doing the even/odd masking at load time (but that doubles the amount of loads, which may cause other problems).